dango/account-margin/src/query.rs

```diff
@@ -1,16 +1,86 @@
 use {
+    anyhow::bail,
     dango_auth::NEXT_SEQUENCE,
-    dango_types::account::single::QueryMsg,
-    grug::{ImmutableCtx, Json, JsonSerExt, StdResult, Storage},
+    dango_lending::DEBTS,
+    dango_oracle::OracleQueries,
+    dango_types::{
+        account::margin::{CollateralPower, HealthResponse, QueryMsg},
+        config::AppConfig,
+    },
+    grug::{
+        Addr, BorshDeExt, Coins, Denom, ImmutableCtx, Inner, IsZero, Json, JsonSerExt, NumberConst,
+        QuerierWrapper, StdResult, Storage, Udec128,
+    },
+    std::collections::BTreeMap,
 };
 
 #[cfg_attr(not(feature = "library"), grug::export)]
-pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> StdResult<Json> {
+pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> anyhow::Result<Json> {
     match msg {
         QueryMsg::Sequence {} => query_sequence(ctx.storage)?.to_json_value(),
+        QueryMsg::Health {} => query_health(ctx)?.to_json_value(),
     }
+    .map_err(Into::into)
 }
 
 fn query_sequence(storage: &dyn Storage) -> StdResult<u32> {
     NEXT_SEQUENCE.current(storage)
 }
+
+/// Calculates the health of a margin account.
+pub fn calculate_account_health(
+    querier: &QuerierWrapper,
+    margin_account: Addr,
+    debts: Coins,
+    collateral_powers: BTreeMap<Denom, CollateralPower>,
+) -> anyhow::Result<HealthResponse> {
```

Comment:
Two optimizations can be done here:

1. Cache prices: if a denom is both a debt and a collateral, we shouldn't query the price twice.
2. If the account doesn't have any debt, then it's healthy for sure. No need to compute the adjusted collateral value.

Let's do these in a later PR.

Reply:
Opened PRs #331 and #332 for this